Source/JavaScriptCore/runtime/VMManager.cpp

```diff
@@ -107,18 +106,426 @@ VM* VMManager::findMatchingVMImpl(const ScopedLambda<VMManager::TestCallback>& t
+            // 1. No targetVM thread was specified (therefore, any thread may service this stop)
+            //    and this is the last thread that stopped. Or ...
+            // 2. This is a subsequent iteration through this loop after context switches (see the
+            //    m_worldConditionVariable.notifyAll() at the bottom of the loop). In which case,
+            //    the targetVM thread is the only one that can get past the wait() above. Or ...
+            // 3. We're executing in RunOne mode and entering this function due to a subsequent
+            //    stop request. In that case, all other threads remained stopped, and only the
+            //    targetVM thread is allowed to run.
+            RELEASE_ASSERT(!m_targetVM || m_targetVM == &vm);
+
+            // Now we can break out of the handler loop is there are no more requests.
+            if (m_currentStopReason == StopReason::None) {
+                if (m_useRunOneMode) {
+                    m_worldMode = Mode::RunOne;
+                    RELEASE_ASSERT(m_targetVM);
+                } else if (m_worldMode != Mode::RunAll)
+                    resumeTheWorld(); // Sets m_worldMode = Mode::RunAll.
+                break; // Exit this loop.
+            }
+
+            m_targetVM = &vm;
+            m_worldMode = Mode::Stopped;
+        }
+
+        auto status = STW_RESUME();
+        switch (m_currentStopReason) {
+        case StopReason::GC:
+            RELEASE_ASSERT_NOT_REACHED();
+        case StopReason::WasmDebugger:
+            status = g_jscConfig.wasmDebuggerStopTheWorld(vm, event);
+            break;
+        case StopReason::MemoryDebugger:
+            status = g_jscConfig.memoryDebuggerStopTheWorld(vm, event);
+            break;
+        case StopReason::None:
+            RELEASE_ASSERT_NOT_REACHED();
+        }
+
+        if (status.first == IterationStatus::Done) {
+            // Done servicing this request. We can't just exit the loop here yet because there
+            // may be other requests that need to be serviced. So, we'll just clear the
+            // current request and go back to the top of the loop to check if there are other
+            // requests. It's safe to clear m_currentStopReason without acquiring m_worldLock
+            // here because currently, all other VM threads are already stopped.
+            // Same reason for why it's safe to set m_useRunOneMode here.
+            auto requestBits = static_cast<StopRequestBits>(m_currentStopReason);
+            m_pendingStopRequestBits.exchangeAnd(~requestBits);
+            m_currentStopReason = StopReason::None;
+
+            // targetVM not being specified means that we should not change m_useRunOneMode.
+            if (status.second) {
+                if (status.second == STW_RESUME_ALL_TOKEN)
+                    m_useRunOneMode = false; // Clearing to ensure we STW_RESUME_ALL.
+                else if (status.second)
```

Comment:
Should be `else`?

Reply:
You are correct.  This if + if else was residual from old logic before I added the outer `if (status.second)`.  It can now be simplified.  I'll make the change.